internal/metrics/collector.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 Aerospike, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"context"
+	"log/slog"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	MetricRecordsPerSecond   = "rps"
+	MetricKilobytesPerSecond = "kbps"
+)
+
+// Collector tracks and logs metrics such as request rate and counts within a context-managed environment.
+type Collector struct {
+	ctx     context.Context
+	enabled bool
+	name    string
+
+	Increment func()
+	Add       func(n uint64)
+
+	counter  atomic.Uint64
+	lastTime time.Time
+
+	lastResultMu sync.RWMutex
+	lastResult   float64
+
+	logger *slog.Logger
+}
+
+// NewCollector initializes a new Collector with the provided context and logger,
+// Enabled metrics will be reported to logger debug level.
+func NewCollector(ctx context.Context, logger *slog.Logger, name string, enabled bool) *Collector {
+	mc := &Collector{
+		ctx:       ctx,
+		enabled:   enabled,
+		name:      name,
+		Increment: func() {},
+		Add:       func(uint64) {},
+		lastTime:  time.Now(),
+		logger:    logger,
+	}
+
+	logger.Debug("metrics", slog.String("name", name), slog.Bool("enabled", enabled))
+	mc.Increment()
+
+	if enabled {
+		mc.Increment = func() { mc.counter.Add(1) }
+		mc.Add = func(n uint64) { mc.counter.Add(n) }
+
+		go mc.report()
+	}
+
+	return mc
+}
+
+// report periodically calculates and logs requests/records per second based on tracked request counts and elapsed time.
+// It operates until the context of the Collector is canceled or its Done channel is closed.
+func (mc *Collector) report() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case t := <-ticker.C:
+			count := mc.counter.Swap(0)
+			elapsed := t.Sub(mc.lastTime).Seconds()
+			rps := float64(count) / elapsed
+
+			// for kbps metric we should divide by 1024.
+			if mc.name == MetricKilobytesPerSecond {
+				rps /= 1024
+			}
+
+			mc.lastResultMu.Lock()
+			mc.lastResult = rps
+			mc.lastResultMu.Unlock()
+
+			mc.logger.Debug("metrics, per second collector", slog.Float64(mc.name, rps))
```

Comment:
It would be useful to have the operation type in the log (currently there is no unique id associated with each operation). Something like:
```suggestion
			mc.logger.Debug("backup metrics collector", slog.Float64(mc.name, rps))
```

Reply:
This metrics is also used for restore, so I tried to make unified log string. And we can identify them by `mc.name`

Reply:
You're right. `backup` was used as an example, it could have been `restore` as well. Without a unique identifier we may have mixed metric logs for several processes running simultaneously. Check if we can add a backup/restore marker, at least. If not, leave the message as `metrics collector` for now.

Reply:
I can add one more param to constructor, to identify